testsuite/MDAnalysisTests/core/test_universe.py

```diff
@@ -1538,3 +1540,308 @@ def test_only_top(self):
+        # [BEFORE] 1 segment in itp
+        assert len(itp.segments) == 1
+        assert_equal(["SOL"], itp.segments.segids)
+        assert len(itp.residues) == 1
+        assert_equal([1], itp.residues.resids)
+        assert_equal([1], itp.residues.chargegroups)
+        assert len(itp.atoms) == 5
+        assert_equal([1] * 5, itp.atoms.resids)
+        assert_equal([1] * 5, itp.atoms.chargegroups)
+        original_attrs = itp._topology.attrs
+
+        # [ACT] set new segids and resids
+        itp.set_groups(
+            atomwise_segids=["A", "A", "B", "B", "B"],
+            atomwise_resids=[3, 3, 3, 3, 4],
+        )
+
+        # [AFTER] 2 segments in itp
+        assert len(itp.segments) == 2
+        assert_equal(["A", "B"], itp.segments.segids)
+        assert len(itp.residues) == 3
+        assert_equal(["A", "B", "B"], itp.residues.segids)
+        assert_equal([3, 3, 4], itp.residues.resids)
+        assert_equal([1, 1, 1], itp.residues.chargegroups)
+        assert len(itp.atoms) == 5
+        assert_equal([3, 3, 3, 3, 4], itp.atoms.resids)
+        assert_equal([1] * 5, itp.atoms.chargegroups)
+        assert len(itp._topology.attrs) == len(original_attrs)
+        assert_equal(
+            sorted([each.attrname for each in itp._topology.attrs]),
+            sorted([each.attrname for each in original_attrs]),
+        )
+
+    def test_with_TPR(self, tpr):
+        # [BEFORE]
+        assert len(tpr.segments) == 3
+        assert_equal(
+            ["seg_0_AKeco", "seg_1_SOL", "seg_2_NA+"], tpr.segments.segids
+        )
+        assert len(tpr.residues) == 11302
+        assert_equal(range(1, 11303, 1), tpr.residues.resids)
+        orignial_res_molnums = tpr.residues.molnums
+        assert len(tpr.atoms) == 47681
+        orignial_resids = tpr.atoms.resids
+        orignial_atm_molnums = tpr.atoms.molnums
+        original_attrs = tpr._topology.attrs
+
+        # [ACT] set new segids and resids
+        sub_dict = {"seg_0_AKeco": "A", "seg_1_SOL": "B", "seg_2_NA+": "C"}
+        new_segids = [sub_dict[each] for each in tpr.atoms.segids]
+        tpr.set_groups(
+            atomwise_segids=new_segids,
+            atomwise_resids=tpr.atoms.resids + 120,
+        )
+
+        # [AFTER]
+        assert len(tpr.segments) == 3
+        assert_equal(["A", "B", "C"], tpr.segments.segids)
+        assert len(tpr.residues) == 11302
+        assert_equal(range(1 + 120, 11303 + 120, 1), tpr.residues.resids)
+        assert_equal(orignial_res_molnums, tpr.residues.molnums)
+        assert len(tpr.atoms) == 47681
+        assert_equal(orignial_resids + 120, tpr.atoms.resids)
+        assert_equal(orignial_atm_molnums, tpr.atoms.molnums)
+        assert len(tpr._topology.attrs) == len(original_attrs)
+        assert_equal(
+            sorted([each.attrname for each in tpr._topology.attrs]),
+            sorted([each.attrname for each in original_attrs]),
+        )
+
+    def test_no_affect_custom_attrs(self, good):
```

Comment:
Great idea to add this test and you checked everything very thoroughly and meticulously -- thank you!!